### PR TITLE
GH-38655: [C++] "iso_calendar" kernel returns incorrect results for array length > 32

### LINKED ISSUE #38655: [C++] "iso_calendar" kernel returns incorrect results for array length > 32
### Describe the bug, including details regarding any error messages, version, and platform.

```python
In [1]: import pyarrow as pa

In [2]: pa.__version__
Out[2]: '14.0.0'

In [3]: import pyarrow.compute as pc

In [4]: from datetime import  datetime

In [5]: arr = pa.array([datetime(2019, 1, 3, 5, 11)]*50)

In [6]: pc.iso_calendar(arr)
Out[6]: 
<pyarrow.lib.StructArray object at 0x1500ff640>
-- is_valid: all not null
-- child 0 type: int64
  [
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    ...
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ]
-- child 1 type: int64
  [
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    2019,
    ...
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ]
-- child 2 type: int64
  [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    ...
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ]

In [7]: pc.iso_calendar(arr)
zsh: segmentation fault  ipython
```

### Component(s)

Python

### PR DESCRIPTION
**Pull request by rok**
### Rationale for this change

When defining `StructArray`'s field builders for `ISOCalendar` we don't pre-allocate memory and then use unsafe append. This causes the resulting array to be at most 32 rows long.

### What changes are included in this PR?

This introduces required memory pre-allocation in the `ISOCalendar` c++ kernel.

### Are these changes tested?

This adds a test for the Python wrapper.

### Are there any user-facing changes?

Fixes the behavior of `iso_calendar` kernel.
* Closes: #38655

### CODE REVIEW

**Comment by rok** [comment]
@jorisvandenbossche I think this is a pretty simple fix.

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc [hunk 1]
```diff
@@ -1503,14 +1503,13 @@ struct ISOCalendar {
     std::unique_ptr<ArrayBuilder> array_builder;
     RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), IsoCalendarType(), &array_builder));
     StructBuilder* struct_builder = checked_cast<StructBuilder*>(array_builder.get());
-    RETURN_NOT_OK(struct_builder->Reserve(in.length));
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc:1506]: Should this be kept? (eg `struct_builder->AppendNull()` is used, which might also set the top-level bitmap?)

File: python/pyarrow/tests/test_compute.py [hunk 2]
```diff
@@ -2223,6 +2223,29 @@ def _check_datetime_components(timestamps, timezone=None):
         first_week_is_fully_in_year=False)
     assert pc.week(tsa, options=week_options).equals(pa.array(iso_week))
 
+    # Test for iso_calendar regression
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_compute.py:2226]: Can you move this to a separate test? The `_check_datetime_components` function is already quite long, and the below also doesn't actually use `timestamps`, so it's a bit strange to put it here)

Comment (rok) [reply python/pyarrow/tests/test_compute.py:2228]: Fair point! Moved to `_test_iso_calendar_regression`.

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc]: I don't know how the top-level bitmap behavior here would be. Reverting the change.

**Comment by rok** [comment]
ping @jorisvandenbossche :)

**Comment by jorisvandenbossche** [comment]
I pushed a small update to simplify the test a bit further (all the test code also needs to be maintained, and for this extra test I don't think it's needed to care about the pandas compatibility etc)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
The one python failure is a know issue, the C++ mac failure is unrelated (s3fs)

**Comment by jorisvandenbossche** [comment]
Thanks!

**Comment by rok** [comment]
Thanks for the review @jorisvandenbossche ! :)

**Pull request merged**